tensorflow/core/common_runtime/pending_counts.h

```diff
@@ -250,6 +263,17 @@ class PendingCounts {
     return AdjustResult(c->dead_count, c->pending);
   }
 
+  template <typename T>
+  inline AdjustResult adjust_for_activation_shared_atomic(T* c, bool increment_dead) {
+    auto* atomic_c = reinterpret_cast<std::atomic<T>*>(c);
```

Comment:
I guess that's true. I did this because the alternative is to make it std::atomic always, and use std::memory_order_relaxed for the non-atomic accesses. Unfortunately that still causes "compiler barriers" (ie prevents reordering/coalescing of writes by the compiler) even though those aren't necessary in the non-atomic case (thread safety is provided by the locks in that case). I'll make that change, though.

Reply:
@toddlipcon, Any update on this PR? Please. Thanks!

Reply:
Sure, happy to make the change if that's the only comment on the review. I was waiting to see if @rmlarsen was going to add comments so I could address them all together. Otherwise I'll make the change in the next day or two

Reply:
I made the change so that the pending count structures are only accessed via std::atomic<T> wrappers, and std::memory_order_relaxed is used for the case where the access is protected by a lock. That said, I also looked at the generated assembly, and it seems that there's quite a bit more code generated now for simple cases like decrementing the pending count. I tested it out on [compiler explorer](https://godbolt.org/z/KqYKnY) and got the following for the old non-atomic code:

```
Adjust(LargeCounts*):
  sub DWORD PTR [rdi], 1
  ret
```

vs for the new code:
```
AdjustAtomic(std::atomic<LargeCounts>*):
  movabs rcx, -4294967296
  mov rax, QWORD PTR [rdi]
  lea edx, [rax-1]
  and rax, rcx
  or rax, rdx
  mov QWORD PTR [rdi], rax
  ret
```

It seems with the atomic load, modify, and store (even with "relaxed" memory order), it's going through the effort of loading the whole 64-bit structure into a register, then copying out the 32-bit pending count, decrementing it, then loading it back into the upper half of the 64-bit register, and storing it back. I haven't had a chance to run benchmarks yet but wanted to check if you still want to go this route despite the blow-up in code size here in this hot path.

It's also worth noting that compiling with clang and libstdc++ doesn't seem to specialize atomic<LargeCounts> to decay it to simple loads but rather calls out to some runtime function __atomic_load and __atomic_store. I'm not sure if TensorFlow supports libstdc++ or mandates a particular standard library.

An alternative is to accept the fact that the reinterpret_cast is technically undefined behavior, rely on a static_assert that the sizeof the atomic is the same as the size of the underlying structure, and ensure that nothing breaks by the fact that this code path is exceedingly well covered by tests. It's unfortunate that http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4013.html was never adopted to provide a standards-compliant way of achieving this.